java/driver/flight-sql/src/main/java/org/apache/arrow/adbc/driver/flightsql/ObjectMetadataBuilder.java

```diff
@@ -0,0 +1,239 @@
+  private final String tableNamePattern;
+  private final String[] tableTypes;
+  private final String columnNamePattern;
+
+  ObjectMetadataBuilder(
+      BufferAllocator allocator,
+      FlightSqlClient client,
+      final AdbcConnection.GetObjectsDepth depth,
+      final String catalogPattern,
+      final String dbSchemaPattern,
+      final String tableNamePattern,
+      final String[] tableTypes,
+      final String columnNamePattern) {
+    this.allocator = allocator;
+    this.client = client;
+    this.depth = depth;
+    this.catalogPattern = catalogPattern;
+    this.dbSchemaPattern = dbSchemaPattern;
+    this.tableNamePattern = tableNamePattern;
+    this.tableTypes = tableTypes;
+    this.columnNamePattern = columnNamePattern;
+    this.root = VectorSchemaRoot.create(StandardSchemas.GET_OBJECTS_SCHEMA, allocator);
+    this.adbcCatalogNames = (VarCharVector) root.getVector(0);
+    this.adbcCatalogDbSchemasWriter = ((ListVector) root.getVector(1)).getWriter();
+    this.adbcCatalogDbSchemasStructWriter = adbcCatalogDbSchemasWriter.struct();
+    this.adbcCatalogDbSchemaTablesWriter =
+        adbcCatalogDbSchemasStructWriter.list("db_schema_tables");
+    this.adbcCatalogDbSchemaNameWriter = adbcCatalogDbSchemasStructWriter.varChar("db_schema_name");
+    this.adbcTablesStructWriter = adbcCatalogDbSchemaTablesWriter.struct();
+    this.adbcTableNameWriter = adbcTablesStructWriter.varChar("table_name");
+    this.adbcTableTypeWriter = adbcTablesStructWriter.varChar("table_type");
+    this.adbcTableColumnsWriter = adbcTablesStructWriter.list("table_columns");
+  }
+
+  private void writeVarChar(VarCharWriter writer, String value) {
+    byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
+    try (ArrowBuf tempBuf = allocator.buffer(bytes.length)) {
+      tempBuf.setBytes(0, bytes, 0, bytes.length);
+      writer.writeVarChar(0, bytes.length, tempBuf);
+    }
+  }
+
+  private boolean patternMatched(String name, String pattern) {
+    if (pattern == null) {
+      return true;
+    }
+
+    return name.matches(pattern.replace("_", ".").replace("%", ".*"));
+  }
+
+  VectorSchemaRoot build() throws AdbcException {
+    // TODO Catalogs and schemas that don't contain tables are being left out
+    FlightInfo info;
+    if (depth == AdbcConnection.GetObjectsDepth.CATALOGS) {
+      info = client.getCatalogs();
+    } else if (depth == AdbcConnection.GetObjectsDepth.DB_SCHEMAS) {
+      info = client.getSchemas(null, dbSchemaPattern);
+    } else {
+      info =
+          client.getTables(
+              null, // TODO pattern match later during processing
+              dbSchemaPattern,
+              tableNamePattern,
+              tableTypes == null ? null : Arrays.asList(tableTypes),
+              depth == AdbcConnection.GetObjectsDepth.ALL);
+    }
+
+    byte[] lastCatalogAdded = null;
+    byte[] lastDbSchemaAdded = null;
+    int catalogIndex = 0;
+
+    for (FlightEndpoint endpoint : info.getEndpoints()) {
```

Comment:
What other drivers have done is factor out the reader logic from the statement so that it doesn't need to be replicated here. Though maybe it's simple enough that it isn't a big deal.